packages/babel-preset-env/src/index.js

```diff
@@ -124,16 +128,37 @@ export default declare((api, opts) => {
   const plugins = [];
   const pluginUseBuiltIns = useBuiltIns !== false;
 
-  if (
-    modules !== false &&
-    moduleTransformations[modules] &&
+  if (modules !== false && moduleTransformations[modules]) {
     // TODO: Remove the 'api.caller' check eventually. Just here to prevent
     // unnecessary breakage in the short term for users on older betas/RCs.
-    (modules !== "auto" || !api.caller || !api.caller(supportsStaticESM))
-  ) {
-    // NOTE: not giving spec here yet to avoid compatibility issues when
-    // transform-modules-commonjs gets its spec mode
-    plugins.push([getPlugin(moduleTransformations[modules]), { loose }]);
+    const shouldTransformESM =
+      modules !== "auto" || !api.caller || !api.caller(supportsStaticESM);
+    const shouldTransformDynamicImport =
+      modules !== "auto" || !api.caller || !api.caller(supportsDynamicImport);
+
+    if (shouldTransformESM) {
+      // NOTE: not giving spec here yet to avoid compatibility issues when
+      // transform-modules-commonjs gets its spec mode
+      plugins.push([getPlugin(moduleTransformations[modules]), { loose }]);
+    }
+
+    if (
+      shouldTransformDynamicImport &&
+      shouldTransformESM &&
+      modules !== "umd"
+    ) {
+      plugins.push([getPlugin("proposal-dynamic-import"), { loose }]);
+    } else {
+      if (shouldTransformDynamicImport) {
+        console.warn(
+          "Dynamic import can only be supported when transforming ES modules" +
+            " to amd, commonjs or systemjs. Only the parser plugin will be enabled.",
```

Comment:
Nit: should we use proper capitalization here (AMD, CommonJS, SystemJS)?

Reply:
```suggestion
            " to AMD, CommonJS or SystemJS. Only the parser plugin will be enabled.",
```